### ISSUE TITLE
Timeouts at seemingly random moments

### ISSUE THREAD

**Issue by Arsanian** [johnwmillr/LyricsGenius#121]
I'm trying to download a huge number of lyrics for a university project. I have files that represent a genre which contain 50 artists I want to download all lyrics from.

So I wrote a python script that scans the folder and reads the lists one by one, trying to download the lyrics for every artist in these lists.

Sometimes the following happens:

Timeout raised and caught:
HTTPSConnectionPool(host='api.genius.com', port=443): Read timed out. (read timeout=5)
Traceback (most recent call last):
  File "lyricsapi.py", line 54, in <module>
    artist = api.search_artist(a.strip(), max_songs=max_songs, sort="title")
  File "/home/duke/anaconda3/envs/dynamusic/lib/python3.7/site-packages/lyricsgenius/api.py", line 356, in search_artist
    song = Song(info, lyrics)
  File "/home/duke/anaconda3/envs/dynamusic/lib/python3.7/site-packages/lyricsgenius/song.py", line 26, in __init__
    self._body = json_dict['song'] if 'song' in json_dict else json_dict
TypeError: argument of type 'NoneType' is not iterable

This error happens pretty randomly, sometimes after 50 texts, sometimes after 600. Earlier today it happened after downloading 113 texts by Eminem, but in the next try it managed to download all 490 of his songs, just to fail after a few songs from the next artist in line.

This also happened, when I ran the script on my server, which has a separate internet connection.

**Version info**
 - Package version 1.7.0
 - OS: Ubuntu 19.10 (also happened on a 18.04 machine)

**Comment by mxdillon** [comment]
I'm facing the same issue

**Comment by GiorgioGhisotti** [comment]
A workaround for this is to use a try...except block and place the request in a while loop
```python
artists = []
while True:
    try:
        artists.append(genius.search_artist(artist, max_songs=10000))
        break
    except:
        pass
```
This will simply retry the call until it works. I successfully used this to scrape the full discography of 50 artists and I didn't run into any further problems.

**Comment by dmlunde** [comment]
@Arsanian how did you manage to narrow down the Eminem number of songs to 490?

**Comment by danielhorizon** [comment]
I've tried the above and am still getting a timeout.. 

`"HTTPSConnectionPool(host='api.genius.com', port=443): Read timed out. (read timeout=5)" `

Any suggestions? I've tried using a timeout as well (for 60 seconds) and tried the while() and a try/catch.

**Comment by ArinkB** [comment]
I am also having this same issue, my loop is pulling lyrics based on the artist name and song title.  then appending that to a list. I have a try and except and the error still pops up. I also have time.sleep(15) just in case. 
The code can run anywhere from 30min - 5hours. It requires a lot of time monitoring.

**Comment by Allerter** [comment]
@ArinkB, could you please provide the following info so we can re-create and debug your issue:
- your traceback
- a minimal working script so that we can re-create the error.

**Comment by ArinkB** [comment]
> @ArinkB, could you please provide the following info so we can re-create and debug your issue:
> 
> * the version of LyricsGenius
> * your traceback
> * a minimal working script so that we can re-create the error.

sure, the dataframe:
![image](https://user-images.githubusercontent.com/54337181/95230906-7b68f480-07d0-11eb-99b5-0fe92149e1d4.png)

`
lyrics = []

def get_lyrics(): #no arguments needed
    while len(lyrics) != len(end_df): 
        genius = lyricsgenius.Genius("API KEY") # call to lyricsgenius
        for track in end_df.values: 
            song = genius.search_song(track[2], track[0])
            try:    
                lyrics.append(song.lyrics) 
            except:
                lyrics.append(np.NAN) 
        time.sleep(40)

       `

The error:
timeout                                   Traceback (most recent call last)
D:\Anaconda\lib\site-packages\urllib3\connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    425                     # Otherwise it looks like a bug in the code.
--> 426                     six.raise_from(e, None)
    427         except (SocketTimeout, BaseSSLError, SocketError) as e:

D:\Anaconda\lib\site-packages\urllib3\packages\six.py in raise_from(value, from_value)

D:\Anaconda\lib\site-packages\urllib3\connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    420                 try:
--> 421                     httplib_response = conn.getresponse()
    422                 except BaseException as e:

D:\Anaconda\lib\http\client.py in getresponse(self)
   1331             try:
-> 1332                 response.begin()
   1333             except ConnectionError:

D:\Anaconda\lib\http\client.py in begin(self)
    302         while True:
--> 303             version, status, reason = self._read_status()
    304             if status != CONTINUE:

D:\Anaconda\lib\http\client.py in _read_status(self)
    263     def _read_status(self):
--> 264         line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
    265         if len(line) > _MAXLINE:

D:\Anaconda\lib\socket.py in readinto(self, b)
    668             try:
--> 669                 return self._sock.recv_into(b)
    670             except timeout:

D:\Anaconda\lib\ssl.py in recv_into(self, buffer, nbytes, flags)
   1240                   self.__class__)
-> 1241             return self.read(nbytes, buffer)
   1242         else:

D:\Anaconda\lib\ssl.py in read(self, len, buffer)
   1098             if buffer is not None:
-> 1099                 return self._sslobj.read(len, buffer)
   1100             else:

timeout: The read operation timed out

During handling of the above exception, another exception occurred:

ReadTimeoutError                          Traceback (most recent call last)
D:\Anaconda\lib\site-packages\requests\adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    438             if not chunked:
--> 439                 resp = conn.urlopen(
    440                     method=request.method,

D:\Anaconda\lib\site-packages\urllib3\connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    723 
--> 724             retries = retries.increment(
    725                 method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]

D:\Anaconda\lib\site-packages\urllib3\util\retry.py in increment(self, method, url, response, error, _pool, _stacktrace)
    402             if read is False or not self._is_method_retryable(method):
--> 403                 raise six.reraise(type(error), error, _stacktrace)
    404             elif read is not None:

D:\Anaconda\lib\site-packages\urllib3\packages\six.py in reraise(tp, value, tb)
    734                 raise value.with_traceback(tb)
--> 735             raise value
    736         finally:

D:\Anaconda\lib\site-packages\urllib3\connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    669             # Make the request on the httplib connection object.
--> 670             httplib_response = self._make_request(
    671                 conn,

D:\Anaconda\lib\site-packages\urllib3\connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    427         except (SocketTimeout, BaseSSLError, SocketError) as e:
--> 428             self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
    429             raise

D:\Anaconda\lib\site-packages\urllib3\connectionpool.py in _raise_timeout(self, err, url, timeout_value)
    334         if isinstance(err, SocketTimeout):
--> 335             raise ReadTimeoutError(
    336                 self, url, "Read timed out. (read timeout=%s)" % timeout_value

ReadTimeoutError: HTTPSConnectionPool(host='api.genius.com', port=443): Read timed out. (read timeout=5)

During handling of the above exception, another exception occurred:

ReadTimeout                               Traceback (most recent call last)
D:\Anaconda\lib\site-packages\lyricsgenius\api\base.py in _make_request(self, path, method, params_, public_api, **kwargs)
     52         try:
---> 53             response = self._session.request(method, uri,
     54                                              timeout=self.timeout,

D:\Anaconda\lib\site-packages\requests\sessions.py in request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
    529         send_kwargs.update(settings)
--> 530         resp = self.send(prep, **send_kwargs)
    531 

D:\Anaconda\lib\site-packages\requests\sessions.py in send(self, request, **kwargs)
    642         # Send the request
--> 643         r = adapter.send(request, **kwargs)
    644 

D:\Anaconda\lib\site-packages\requests\adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    528             elif isinstance(e, ReadTimeoutError):
--> 529                 raise ReadTimeout(e, request=request)
    530             else:

ReadTimeout: HTTPSConnectionPool(host='api.genius.com', port=443): Read timed out. (read timeout=5)

During handling of the above exception, another exception occurred:

Timeout                                   Traceback (most recent call last)
<ipython-input-47-3ed120674112> in <module>
----> 1 get_lyrics() # calling the function

<ipython-input-46-c755c4d9aebe> in get_lyrics()
      7         genius = lyricsgenius.Genius("2AaY5B5Iylvf3X2oQl_4Jmk74pSZ-J8VqcboMk1PO_y7N8wlN83D9a85uhNh49Id") # call to lyricsgenius
      8         for track in end_df.values[7729:]: # for each track in the dataframe. I have indexed it to 5 rows for submission only
----> 9             song = genius.search_song(track[1], track[0]) # find song on Genius with matching name and artist
     10             try:
     11                 l2.append(song.lyrics) # append value to list

D:\Anaconda\lib\site-packages\lyricsgenius\genius.py in search_song(self, title, artist, get_full_info)
    511         song_info = result.copy()
    512         if get_full_info:
--> 513             song_info.update(self.song(result['id'])['song'])
    514         lyrics = self.lyrics(song_info['url'])
    515 

D:\Anaconda\lib\site-packages\lyricsgenius\genius.py in song(self, song_id, text_format, public_api)
    419             return super(PublicAPI, self).song(song_id, text_format)
    420         else:
--> 421             return super().song(song_id, text_format)
    422 
    423     def search_songs(self, search_term, per_page=None, page=None, public_api=False):

D:\Anaconda\lib\site-packages\lyricsgenius\api\api.py in song(self, song_id, text_format)
    424         endpoint = "songs/{id}".format(id=song_id)
    425         params = {'text_format': text_format or self.response_format}
--> 426         return self._make_request(endpoint, params_=params)
    427 
    428     def web_page(self, raw_annotatable_url=None, canonical_url=None, og_url=None):

D:\Anaconda\lib\site-packages\lyricsgenius\api\base.py in _make_request(self, path, method, params_, public_api, **kwargs)
     58         except Timeout as e:
     59             error = "Request timed out:\n{e}".format(e=e)
---> 60             raise Timeout(error)
     61         except HTTPError as e:
     62             error = str(e)

Timeout: Request timed out:
HTTPSConnectionPool(host='api.genius.com', port=443): Read timed out. (read timeout=5)

**Comment by Allerter** [comment]
@ArinkB, thanks for providing the information. Although this issue is probably a valid issue, I don't think your script's primary issue is the one with the Timeout. I tested Spotify's Viral 50 songs using your script and here are a couple of things that you could improve:
```python
from requests.exceptions import Timeout

lyrics = []


def get_lyrics():
    # while len(lyrics) != len(end_df): #1
    genius = lyricsgenius.Genius(token)
    genius.timeout = 15
    genius.sleep_time = 40  # 2
    # or: Genius(token, timeout=15, sleep_time=40)
    for track in end_df.values:
        retries = 0
        while retries < 3:
            try:
                song = genius.search_song(track[2], track[0])
            except Timeout as e:
                retries += 1
                continue
            if song is not None:
                lyrics.append(song.lyrics)
            else:
                lyrics.append(np.NAN)
            break
```

1. This will result in an infinite loop since some songs can't be found, and there's no need for it in the first place.
2. With the `genius.sleep_time` attribute, there's no need for `time.sleep(40)` anymore. Also, I don't think there's a need for a 40-sec sleep from the API's end. When I tested your script, I removed the `time.sleep(40)` line and everything worked fine.

Now your script will search for the songs and in case of timeouts, your script will retry the search three times before moving on to the next song (this should probably be a feature, @johnwmillr).

**Comment by ArinkB** [comment]
@Allerter Thank you! I appreciate your help and insight. It has been pulling for 3 hours now and no issues so far.

**Comment by NIkitabala** [comment]
@ArinkB Hi, can you show me, how exactly do you use your script? I'm trying to use this solution, but I'm still getting an error.

**Comment by ArinkB** [comment]
@NIkitabala sure,
this is the notebook I used it in, I modified it slightly because my original project plan didn't work out at the time:
https://github.com/ArinkB/Predicting-Song-Skips/blob/master/1_Data%20Acquisition.ipynb

**Comment by Allerter** [comment]
Based on [this comment](https://github.com/johnwmillr/LyricsGenius/issues/168#issuecomment-720739743) that I posted on #168, I think these random timeout errors will be solved by #162. We'll see.
